### DISCUSSION TITLE
RFC: Migrate from CJS to ESM

### DISCUSSION THREAD

**Discussion by twilson63** [hyper63/atlas#107] [Ideas] [answered]
- 2021-02-05
- RFC PR: (leave this empty)
- hyper63 Issue: (leave this empty)

# Proposal to migrate from CJS to ESM

## Summary

> It is 2021 and the javascript ecosystem needs to move from cjs to esm, node v14 supports native esm, we should be building in esm, but without the transpilation process of babel. 

## Motivation

> Why are we doing this? What use cases does it support? What is the expected
outcome?

We will stay current with the javascript ecosystem and get better code analysis tooling and cross-platform compatibility.

## Drawbacks

> Why should we *not* do this? Please consider the impact on teaching Atlas,
on the integration of this feature with other existing and planned features,
on the impact of the API churn on existing apps, etc.

* dependencies, some dependencies simply won't work with nodejs native esm
* testing, the test runner for tape does not work for esm, but you can call node as the runner so we will have to create a test/index.js file that imports all of the other test files and runs the test, need to prove out this concept.

> There are tradeoffs to choosing any path, please attempt to identify them here.

## Alternatives

> What other designs have been considered? What is the impact of not doing this?

Migrate to deno, deno would give us a platform that has the hindsight of nodejs history, the 
only con to deno is that it does not have a large ecosystem.

> This section could also include prior art, that is, how other frameworks in the same domain have solved this problem.

## Unresolved questions

> Optional, but suggested for first drafts. What parts of the design are still
TBD?

**Comment by TillaTheHun0** [comment, marked as answer]
We migrated to Deno :)
